src/Tracer.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Exercism\PhpTestRunner;
+
+use Exercism\PhpTestRunner\Result;
+use PHPUnit\Event\Code\TestMethod;
+use PHPUnit\Event\Event;
+use PHPUnit\Event\Test\BeforeFirstTestMethodErrored;
+use PHPUnit\Event\Test\Errored;
+use PHPUnit\Event\Test\Failed;
+use PHPUnit\Event\Test\Passed;
+use PHPUnit\Event\Test\PrintedUnexpectedOutput;
+use PHPUnit\Event\TestRunner\Finished;
+use PHPUnit\Event\Tracer\Tracer as TracerInterface;
+use ReflectionClass;
+
+final class Tracer implements TracerInterface
+{
+    private array $result = [
+        'version' => 3,
+        'status' => 'pass',
+        'tests' => [],
+    ];
+
+    public function __construct(
+        private readonly string $outFileName,
+        private readonly string $exerciseDir,
+    ) {
+    }
+
+    public function trace(Event $event): void
+    {
+        match (\get_class($event)) {
+            Passed::class => $this->addTestPassed($event),
+            Failed::class => $this->addTestFailed($event),
+            Errored::class => $this->addTestErrored($event),
+            BeforeFirstTestMethodErrored::class => $this->addBeforeFirstTestMethodErrored($event),
+            PrintedUnexpectedOutput::class => $this->addTestOutput($event),
+            // default => $this->addUnhandledEvent($event),
+            default => true,
+        };
+
+
+        if ($event instanceof Finished) {
+            $this->saveResults();
+        }
+    }
+
+    private function addUnhandledEvent(Event $event): void
+    {
+        $this->result['tests'][] = new Result(
+            $event->asString(),
+            'fail',
+            'no code yet',
+        );
+    }
+
+    private function addTestPassed(Passed $event): void
+    {
+        /** @var TestMethod */
```

Comment:
Should this be converted to an `assert`?

Reply:
I'm not sure about the gain of the `assert`? There is no handling for an `assert` failure that would yield a result we could see. If we don't produce the `result.json` we do not get any feedback to work with. So `assert` does not help with anything a `Method does not exist` error doesn't, too.

Reply:
It just makes the annotation concrete. So is ensure it remains corrects in the future. Maybe the `testDox()` method will still exist on a new object but it won't be the same object.

Reply:
I see that. But without `zend.assertion=1` in php.ini (we have only production setting `-1` when running things) this is very much useless. I add that to #131.